items.md

```diff
@@ -23,37 +23,42 @@ Provide detailed information about a specific item. The JSON response document i
         "value": "Stvilia, Besiki",
         "language": "en",
         "authority": null,
-        "confidence": -1
+        "confidence": -1,
+        "place": 0
```

Comment:
Yes, it was discussed with Chris and introduced to support re-ordering authors in a next phase

Reply:
Can you provide details about how the re-ordering authors is planned to be supported? otherwise it is better to postpone this change to when we will know how it will be used

Reply:
There's no detail about the re-ordering yet, but the place column is also essential because it determines the order when mixing plain-text and entities relationships as explained in
https://docs.google.com/document/d/1X0XsppZYOtPtbmq7yXwmu7FbMAfLxxOCONbw0_rl7jY/edit#heading=h.3z0989mcmfe6

Reply:
I still don't see why the place is required for sure. When you mix plan and relation metadata you will still have a clear order in the json. The client should be unaware of implementation details like the one discussed in the google doc above.
As it is not required by the current implemented features please keep this change out of this PR

Reply:
I'm sorry to hear it's not clear to you yet why the place is important.
Feel free to explain why it's a problem to have it included, so we can discuss it in the working group whether it has to be removed again

Reply:
Similar to `PUT` discussion (see below in this same review), my main question though is whether removing this functionality (from the main code) _now_ will take significant effort. This might be a scenario were we need to consider making an exception just to move this feature along.

Once we get to the next phase, we do have the opportunity to rethink the approach and determine if we want this removed.  In other words, if this is effort that will be significant now, we could create a JIRA ticket to revisit this (as we've been doing with a lot of undecided topics or cleanup steps for Configurable Entities).  My plan is to go through those JIRA tickets in the next meeting of the Entities WG to plot out the next steps & ensure all are addressed prior to 7.0 beta/final.

Reply:
please see the general comment here https://github.com/DSpace/Rest7Contract/pull/57#discussion_r281163146

Reply:
@abollini : As the `place` field has existed for some time in DSpace (at the database layer) and also has a purpose described in the Google doc (which @benbosman linked to above), I'd recommend we allow this in for now.  I can create a JIRA ticket to (later) re-analyze whether there is a way to use the JSON ordering instead of `place`...but that seems to be a bigger discussion/analysis, which would slow down this PR significantly.

Reply:
Here's the JIRA ticket to log this work: https://jira.duraspace.org/browse/DS-4242

Reply:
Hi @tdonohue ,
I would try to restore a correct development process here that follow the agreed rules for DSpace 7.
We are following now a Contract first, test driven dev approach. This mean that code needs to follow the contract not viceversa. We agreed to have a highly uniform REST API, following standards, best and wide adopted best practices and we try to add direction in such sense in our REST contract home page

There is no need to discuss again and again the basic principles (verbs to use, structure of the endpoint, meaning of the payload) for each new endpoint. What I expect is that the basic principles will be always respected and that exceptions will be introduced only  *after discussion and once proven that they are really necessary*. It doesn't work in the opposite way, we don't need to proof that the general way work here. We don't need to discuss about removing later the place attribute, we need to discuss about adding back the place attribute.

The role of the reviewers in the REST contract is to double check that no deviation are introduced, clarify and extend the general rules if needed and only as last chance verify that proof have been discussed justifying a deviation if any. It seems to me that the general principles here provide answers to our need, I haven't hear the opposite. 

I really cannot understand why all this discussion around the request to respect our methodology and change *one line of code*. As the reorder feature is not yet implemented how this can create significant effort? why you want to put in the contract, that set the direction of the development, something that we are saying should be really avoided? Please note that the section in the google doc only give details about how the place information will be internally used to allow reordering of mixed plan and relation metadata from here there is no implication for the REST contract that should be as much agnostic as possible respect to such details.

Drop these changes at least from the contract, if remove stuff that should not exist from the code require to much effort, create a dedicated bug ticket for that. Please use the last ticket that you have created (DS-4242) *to discuss and proof the need* of a deviation from the general rules.

Reply:
@abollini : Please keep in mind your tone in these messages. While I understand you want to give honest feedback, your tone is sometimes coming across as negative.  Negativity in PR reviews can be frustrating to other developers, especially new ones (and, honestly, it doesn't encourage new developers to get involved in the DSpace 7 efforts).

I'm doing my best to ensure all your Entities feedback is taken into account. However, as we all know, the implementation PR (https://github.com/DSpace/DSpace/pull/2376) is *massive* and very complex to maintain/update (and as we found, it seems we are beginning to hit GitHub bugs or limitations, as some comments in the large PR are not appearing on the correct line). I'm also trying to ensure we are staying on schedule to prepare ourselves for the upcoming OR2019 workshops/talks. Therefore, I'd like to find a quick compromise on how we can move this forward quickly while also ensuring outstanding updates/discussions are prioritized for fixing immediately after the merger. As these PRs are both at +2, technically they could be merged. But, I appreciate your expertise and feedback here and am trying to ensure minor tweaks are fixed immediately, while larger changes (or anything needing more discussion) are prioritized for immediate follow-up PRs.

You are correct that the approach of finalizing the REST Contract before implementation is definitely the ideal scenario. However, as we've seen in many (perhaps most) PRs, oftentimes we have to go back and update the Contract to align better with the implementation. In cases like Configurable Entities (where the implementation cannot be easily aligned with existing contract immediately), I'd prefer that the Contract describe the current implementation (instead of a future goal).  That way developers building against the REST API have an accurate Contract to work from.  The contract can always be updated again in the future (as you noted it's one line) once we improve the implementation strategy here.

To try and clarify why the implementation is not easy to align with the Contract in this area...much of this Entities code was implemented *prior* to PR #39, where `place` was first removed from our Contract. The decision in #39 was correct based on the set of use cases at that time (as "place" obviously isn't required by basic metadata responses). However, obviously, Configurable Entities brings in additional use cases which had not been included in that prior decision. Currently, on the Angular side, there are already areas of the Entities code that rely on the existence of the `place` attribute, especially when intermixing entities and plain-text values (in the proper order). In fact, the corresponding Angular PR already has built in tests to verify that Angular is *always trusting the `place` value, and not the ordering of results returned*.  See the `item.component.spec.ts` here: https://github.com/DSpace/dspace-angular/pull/372/files#diff-5b1f0d2025d9a9d276ef0129429d1101R336 and here: https://github.com/DSpace/dspace-angular/pull/372/files#diff-5b1f0d2025d9a9d276ef0129429d1101R409

So, to attempt to change the implementation of the REST API PR would have a snowball effect, in that it will require also rethinking the Angular implementation.  Rather than having this detailed discussion/analysis/rework now, I was recommending we schedule this for revisiting in the Entities WG (via JIRA Ticket DS-4242). It may be possible that we can find a way to address this by (again) removing the `place` attribute. But, I don't want to jump immediately to that conclusion without analysis of upcoming Configurable Entities features.

In summary, if we want the Contract to represent what the Implementation actually does, and we want these PRs moved forward quickly (in order to get to a Preview Release prior to OR2019), then I think we need to consider making an exception to the rule here.  Therefore, I'd like to schedule this analysis as a future task for the Entities WG (in DS-4242). I appreciate all your analysis and feedback on these PRs, but I've definitely been trying to determine which of your requests can be immediately fixed, and which we should schedule for later. I think this one is the latter, for all the reasons described here.

Reply:
Hi @tdonohue ,
I feel that we are making a mistake here and I cannot avoid to tell that.
I understand that rework is frustrating for developers and this is why I stress that we need to respect the rules that we have set for our-self and have a valuable conversation before to write the most of the code. You are right that most of the past PRs haven't followed these rules but if anytime we end saying that we will never improve our processes... anyway, your decision is clear and also if I disagree I can accept it.

Please keep in mind that nobody likes to spent time reviewing code and end up saying things that will be perceived necessary as negative because it is simply too late and we are in the rush due to an ineffective process.

Reply:
All, I've just added a warning notice into the Contract about the existence of the `place` field (and linking back to the DS-4242 ticket about it). See this commit: https://github.com/DSpace/Rest7Contract/pull/57/commits/5db8383010ebfbfc5df3855ee885ff41a8121dba

Reply:
I want to thanks  @mwoodiupui to have explained in a clear way the message that I really feel important: we should not allow this situation to recur in future, I like to discuss in a constructive way, during the design phase, without the pain to request rework. We need to use in the most effective way the resource of our community: developer times. 

We are making changes to our processes here to be more inclusive and raise the quality of our project. Sometimes a developer doesn't have the time to follow all the details of an implementation but struggle to be involved in the key decisions, the contract first approach will help in achieve that. If the current situation doesn't offer us other options I feel that the current notes and tickets are a good compromise. 

**I hope to see all the future PR, related to whatever functionality/change follow the contract first approach.**
The first **rest contract** doesn't need to be perfect or complete, **soon will be better than perfect**. Start a discussion with just a high level idea and several options will be IMHO the ideal scenario.